### PR TITLE
[TE] rootcause - auto populate selection

### PR DESCRIPTION
**Pull request by apucher**
This PR adds a setup mode to /rootcause.

First, this enables the auto-populate use-case when entering from blank, metric, or anomaly state (i.e. all but existing session). After the user selects a metric, the top-ranking entities can now be added automatically to the selection by default based on a score threshold. (this is disabled by default)

Second, this enables distinct default time ranges when selecting the initial metric. Using metric-level information about granularity and data range we can select an appropriate time window.

The PR also makes minor additions to the backend to surface needed information.

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-data-indicator/component.js [hunk 1]
```diff
@@ -27,15 +27,14 @@ export default Component.extend({
   didReceiveAttrs() {
     const { selectedUrn, anomalyRange } = getProperties(this, 'selectedUrn', 'anomalyRange');
 
-    if (!selectedUrn || !anomalyRange) { return; }
+    if (!selectedUrn || !anomalyRange || !selectedUrn.startsWith('thirdeye:metric:')) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/components/rootcause-data-indicator/component.js]: Let's add a comment explaining why this is necessary

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 2]
```diff
@@ -254,14 +290,46 @@ export default Ember.Controller.extend({
...
+        case ROOTCAUSE_SETUP_MODE_CONTEXT:
+          this._setupContext();
+          break;
+
+        case ROOTCAUSE_SETUP_MODE_SELECTED:
+          this._setupSelected();
+          break;
+
+        default:
+          throw new Error(`Unknown setup mode '${setupMode}'`);
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: Is this error currently getting caught?

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 3]
```diff
@@ -530,6 +598,79 @@ export default Ember.Controller.extend({
...
+    const newSelectedUrns = new Set(selectedUrns);
+
+    const contextMetricUrns = filterPrefix(context.urns, 'thirdeye:metric:');
+    contextMetricUrns.forEach(urn => newSelectedUrns.add(urn));
+    contextMetricUrns.map(toCurrentUrn).forEach(urn => newSelectedUrns.add(urn));
+    contextMetricUrns.map(toBaselineUrn).forEach(urn => newSelectedUrns.add(urn));
+
+    // events
+    const groupedEvents = Object.values(entities)
+      .filter(e => e.type === 'event')
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: Let's try using more meaningful variable names than single letters for readability sake. 
The final compile step changes them to be more succinct anyway.

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 4]
```diff
@@ -530,6 +598,79 @@ export default Ember.Controller.extend({
...
+        const topk = arr.filter(e => e.score >= ROOTCAUSE_SETUP_EVENTS_SCORE_THRESHOLD);
+        topk.forEach(e => newSelectedUrns.add(e.urn));
+      });
+
+    // metrics
+    const metricUrns = filterPrefix(Object.keys(entities), 'thirdeye:metric:')
+      .filter(urn => urn in scores)
+      .filter(urn => scores[urn] >= ROOTCAUSE_SETUP_METRICS_SCORE_THRESHOLD);
+
+    metricUrns.forEach(urn => newSelectedUrns.add(urn));
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: qq: any reason why this wasn't chained after the `.filter`?

File: thirdeye/thirdeye-frontend/app/pods/rootcause/route.js [hunk 5]
```diff
@@ -7,17 +7,77 @@ import { toCurrentUrn, toBaselineUrn, filterPrefix, dateFormatFull } from 'third
...
+const ROOTCAUSE_SETUP_MODE_SELECTED = "selected";
+const ROOTCAUSE_SETUP_MODE_NONE = "none";
+
+/**
+ * converts RCA backend granularity strings into units understood by moment.js
+ */
+const toMetricGranularity = (attrGranularity) => {
+  let [count, unit] = attrGranularity.split('_');
+
+  switch (unit) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/route.js]: I personally prefer the shorter way:

```
const UNIT_MAPPING = {
  MINUTES: 'minute',
  HOURS: 'hour',
  DAYS: 'day'
}
then UNIT_MAPPING[unit];
```

This also allows moving this constant to the `constants.js` file (if needed)

File: thirdeye/thirdeye-frontend/app/pods/rootcause/route.js [hunk 6]
```diff
@@ -7,17 +7,77 @@ import { toCurrentUrn, toBaselineUrn, filterPrefix, dateFormatFull } from 'third
...
+      break;
+  }
+
+  return [parseInt(count, 10), unit];
+};
+
+/**
+ * Returns the anomaly time range offset (in granularity units) based on metric granularity
+ */
+const toAnomalyOffset = (granularity) => {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/route.js]: Harde coded arbitrary nums?

File: thirdeye/thirdeye-frontend/app/pods/rootcause/route.js [hunk 7]
```diff
@@ -27,10 +87,11 @@ export default Ember.Route.extend(AuthenticatedRouteMixin, {
   model(params) {
     const { metricId, anomalyId, sessionId } = params;
 
-    let metricUrn, anomalyUrn, session, anomalyContext, anomalySessions;
+    let metricUrn, metricEntity, anomalyUrn, session, anomalyContext, anomalySessions;
 
     if (metricId) {
       metricUrn = `thirdeye:metric:${metricId}`;
+      metricEntity = fetch(`/rootcause/raw?framework=identity&urns=${metricUrn}`).then(checkStatus).then(res => res[0]).catch(() => undefined);
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/route.js]: we can just pass an anonymous function in catch
`.catch(() => {})` 
This essentially achieves the same thing, but looks better imho

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js (hunk 2 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: No. This is a "must not happen" case. We can also ignore it transparently, but may see unexpected behavior later

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js (hunk 3 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: style question. not sure if a long variable in lambda expressions make this more readable. ultimately, we'll have to adopt a code style guideline for our JS

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js (hunk 4 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/rootcause/route.js (hunk 5 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/route.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/rootcause/route.js (hunk 6 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/route.js]: yes. if it needs a config, we can package the logic here too

**Pull request merged**